File: 2001-3000/2354-number-of-excellent-pairs/2354-number-of-excellent-pairs.cpp
class Solution {
public:
    long long countExcellentPairs(vector<int>& nums, int k) {

        sort(nums.begin(), nums.end());
        nums.erase(unique(nums.begin(), nums.end()), nums.end());

        vector<long long> freq(33, 0);

        for (int x : nums) {
            freq[__builtin_popcount(x)]++;
        }

        long long ans = 0;

        for (int i = 0; i <= 32; i++) {
            for (int j = 0; j <= 32; j++) {

                if (i + j >= k) {
                    ans += freq[i] * freq[j];
                }
            }
        }

        return ans;
    }
};

File: 2001-3000/2354-number-of-excellent-pairs/2354-number-of-excellent-pairs.java
class Solution {
    public long countExcellentPairs(int[] nums, int k) {
        Set<Integer> distinct = new HashSet<>();
        int[] bitCount = new int[31];
        for(int i : nums) {
            if(distinct.add(i)) {
                bitCount[Integer.bitCount(i)]++;
            }
        }
        int[] sum = new int[31];
        for(int i = 30; i >= 0; i--) {
            sum[i] = bitCount[i] + (i == 30 ? 0 : sum[i + 1]);
        }
        long ans = 0;
        for(int i = Math.max(1, k - 30); i <= 30; i++) {
            ans += ((long) bitCount[i]) * sum[Math.max(0, k - i)];
        }
        return ans;
    }
}